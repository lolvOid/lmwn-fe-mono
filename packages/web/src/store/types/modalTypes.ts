export type ModalDataType = {
    restaurantId: string;
    menuName: string;
};
export type ModalStore = {
    show: boolean;
    modalData: ModalDataType | undefined;
    showModal: () => void;
    hideModal: () => void;
    setModalData: (value: ModalDataType) => void;
    resetModalData: () => void;
};
