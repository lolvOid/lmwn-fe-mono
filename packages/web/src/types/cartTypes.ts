export type CartDataType = {
    id: number | string;
    name: string | undefined;
    restaurantId: number | string;
    fullPrice: number;
    count: number;
    options: string[];
    thumbnailImage: string | undefined;
};

export type CartStore = {
    isCartDisplayed: boolean;
    cartData: CartDataType[];
    totalCartPrice: number;
    totalCartCount: number;
    showCart: () => void;
    hideCart: () => void;
    addToCart: (value: CartDataType) => void;
    resetCart: () => void;
    removeFromCartById: (id: number | string) => void;
    increaseCount: (id: number | string) => void;
    decreaseCount: (id: number | string) => void;
    updateOptions: (id: number | string, options: string[]) => void;
    updateCartCountById: (id: number | string, count: number) => void;
    calculateTotalCartPrice: () => void;
    calculateTotalCartCount: () => void;
};
