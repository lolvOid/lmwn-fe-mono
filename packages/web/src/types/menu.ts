export interface Menu {
  name: string;
  id: string;
  thumbnailImage?: string;
  fullPrice: number;
  discountedPercent: number;
  discountedTimePeriod?: {
    begin: string;
    end: string;
  };
  sold: number;
  totalInStock: number;
}

export interface FullMenu extends Menu {
  largeImage?: string;
  options: {
    label: string;
    choices: {
      label: string;
    }[];
  }[];
}
