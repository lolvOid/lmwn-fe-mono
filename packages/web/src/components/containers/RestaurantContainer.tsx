import RestaurantCardHeader from '@/components/restaurant-card/RestaurantCardHeader';
import ItemCardLayout from '@/components/layouts/ItemCardLayout';
import ItemCard from '@/components/card/ItemCard';
import { Fragment } from 'react';
import { Menu } from '@/interfaces/menu';
import useModalStore from '@/store/modal/modalStore';
import { getKey } from '@/utils/generateId';
import { RestaurantContainerProps } from './RestaurantContainerProps';
import styles from '@/components/containers/RestaurantContainer.module.scss';
import DotsLoading from '../loading/DotsLoading';

const RestaurantContainer: React.FC<RestaurantContainerProps> = ({
    pageData,
    name,
    openTime,
    closeTime,
    restaurantImage,
    isLoading,
    onLayoutScroll,
}: RestaurantContainerProps) => {
    const { showModal, setModalData } = useModalStore();
    const handleShowItem = (name: string, restaurantId: string) => {
        showModal();
        setModalData({
            menuName: name,
            restaurantId: restaurantId,
        });
    };
    return (
        <main className="w-full max-h-screen relative overflow-y-visible">
            <div className="mx-auto my-0 relative">
                <div className="flex sticky top-16">
                    <div className="w-full mx-auto my-0">
                        <img
                            src={restaurantImage}
                            alt={name}
                            className={styles.containerImage}
                        />
                    </div>
                </div>
                <RestaurantCardHeader name={name} closeTime={closeTime} openTime={openTime} />
                <ItemCardLayout onLayoutScroll={onLayoutScroll}>
                    {pageData &&
                        pageData.pages.length &&
                        pageData.pages.map((page) => {
                            return (
                                <Fragment key={getKey()}>
                                    {page.menus.map((item: Menu, index: number) => {
                                        return (
                                            <ItemCard
                                                key={index}
                                                imageSource={item.thumbnailImage ?? ''}
                                                name={item.name}
                                                fullPrice={item.fullPrice}
                                                discountedPercent={item.discountedPercent}
                                                totalInStock={0}
                                                onClick={() => handleShowItem(item.name, page?.id)}
                                            />
                                        );
                                    })}
                                </Fragment>
                            );
                        })}
                </ItemCardLayout>
                {isLoading && <DotsLoading />}
            </div>
        </main>
    );
};

export default RestaurantContainer;
