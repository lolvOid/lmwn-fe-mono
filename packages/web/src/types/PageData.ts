import { Menu } from '@/interfaces/menu';

export type PageData = {
    pageParams?: number[];
    pages: {
        id: string;
        menus: Menu[];
    }[];
};
