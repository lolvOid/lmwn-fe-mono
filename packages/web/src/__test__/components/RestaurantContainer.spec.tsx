import { render } from '@testing-library/react';
import RestaurantContainer, {
    RestaurantContainerProps,
} from '@/components/containers/RestaurantContainer';
import useModalStore from '@/store/modal/modalStore';
import renderer from 'react-test-renderer';
import { ModalStore } from '@/types/modalTypes';

jest.mock('@/store/modal/modalStore', () => ({
    __esModule: true,
    default: jest.fn(),
}));

describe('RestaurantContainer', () => {
    const mockItemData = {
        pages: [
            {
                id: 'pageId1',
                menus: [
                    {
                        thumbnailImage: 'item1-image.jpg',
                        name: 'Item 1',
                        fullPrice: 200,
                        discountedPercent: '10',
                    },
                ],
            },
        ],
    };

    const mockProps = {
        itemData: mockItemData as unknown,
        name: 'Restaurant Name',
        openTime: '08:00 AM',
        closeTime: '10:00 PM',
        restaurantImage: 'restaurant-image.jpg',
    } as RestaurantContainerProps;

    it('Should render properly with provided props', () => {
        const mockShowModal = jest.fn();
        const mockSetModalData = jest.fn();
        (useModalStore as unknown as jest.Mock<ModalStore>).mockReturnValue({
            show: false,
            modalData: {
                menuName: '',
                restaurantId: '',
            },
            setModalData: mockSetModalData,
            showModal: mockShowModal,
            hideModal: jest.fn(),
            resetModalData: jest.fn(),
        });

        render(<RestaurantContainer {...mockProps} />);
        expect(mockShowModal).not.toHaveBeenCalled();
        expect(mockSetModalData).not.toHaveBeenCalled();

        const containter = renderer.create(<RestaurantContainer {...mockProps} />).toJSON();
        expect(containter).toMatchSnapshot();
    });
});
