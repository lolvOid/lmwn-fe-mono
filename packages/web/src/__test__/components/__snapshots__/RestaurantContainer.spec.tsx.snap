// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantContainer Should render properly with provided props 1`] = `
<main
  className="w-full max-h-screen relative overflow-y-visible"
>
  <div
    className="mx-auto my-0 relative"
  >
    <div
      className="flex sticky top-16"
    >
      <div
        className="w-full mx-auto my-0"
      >
        <img
          alt="Restaurant Name"
          className="containerImage"
          src="restaurant-image.jpg"
        />
      </div>
    </div>
    <div
      className="bg-gray-900 rounded-t-2xl text-white lg:max-w-[600px] -my-3 lg:-my-4 mx-auto w-full px-8 lg:py-4 py-2 sticky z-[400] flex flex-col"
    >
      <span
        className="text-lg"
      >
        Restaurant Name
      </span>
      <span
        className="text-sm text-yellow-300 line"
      >
        Open 
        08:00 AM
         - Close 
        10:00 PM
      </span>
    </div>
    <div
      className="relative flex justify-center items-start overflow-y-auto max-w-[600px] py-8  bg-white lg:min-w-[200px] min-w-full mt-0 mx-auto my-0 z-50 itemCardLayout"
    >
      <div
        className="mx-0 lg:mx-auto my-0 w-full lg:w-auto px-4"
      >
        <div
          className="grid grid-cols-2 lg:grid-cols-2 gap-4 py-2"
        >
          <div
            className="relative flex lg:flex-row flex-col lg:justify-between border-solid border-gray-200 border-[0.4px] p-2 rounded-md opacity-100 cursor-pointer hover:bg-gray-100"
            data-testid="item-card"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="lg:w-32 w-full max-w-40 h-32 flex-shrink-0 justify-center items-center"
            >
              <img
                alt="Item 1"
                className="w-full h-full object-cover object-center"
                src="item1-image.jpg"
              />
            </div>
            <div
              className="px-2 py-2 flex lg:flex-col w-full flex-grow lg:items-end items-start"
            >
              <div
                className="w-full flex flex-col text-sm flex-grow gap-1"
              >
                <span
                  className="text-xs text-wrap"
                  data-testid="item-card-name"
                >
                  Item 1
                </span>
                <span
                  className="text-sm primary-text text-wrap"
                  data-testid="item-card-discounted-price"
                >
                  ฿180
                </span>
                <span
                  className="text-xs text-wrap text-gray-600 line-through"
                  data-testid="item-card-price"
                >
                  ฿200
                </span>
              </div>
              <div
                className="lg:flex hidden"
              >
                <button
                  className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300 hover:animate-pulse"
                  onClick={[Function]}
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                </button>
              </div>
            </div>
            <div
              className="flex lg:hidden absolute bottom-1 z-[200] right-1"
            >
              <button
                className="primary-bg text-sm w-8 h-8 inline-flex items-center justify-center text-white rounded-full hover:bg-yellow-300"
                onClick={[Function]}
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
